students/kyle_lehning/Lesson09/cli_main.py

```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python3
+import tempfile
+import time
+import sys
+import donor_models as dm
+"""
+A  module for handling command line interface for a donor program
+"""
+
+donors = dm.DonorCollection()
+
+
+def thank_you():
+    """
+    Print menu options for writing a thank you letter.
+    If 'list' is provided, launch list_donors function.
+    if 'menu' is provided, this function terminates
+    If a name is provided, launch write_letter function
+    """
+    thank_you_options = {
+        "MENU": False,
+        "LIST": list_donors
+    }
+    while True:
+        user_selection = (
+            input("\nPlease provide a full name to send thank you note to (options 'list' and 'menu'): ")
+        )
+        action = thank_you_options.get(user_selection.upper())
+        if action is False:
+            break
+        elif action is None:  # Write letter to user selection if an action option wasn't selected
+            write_letter(user_selection)
+        else:
+            action()
+
+
+def list_donors():
+    """Go through each donor in public donors object and print to console"""
+    for name in donors.list_donor_names():
+        print(name)
+
+
+def write_letter(person):
+    """
+    Receive the name of a person.
+    Look through people in donors object and use that donor if they exist, if they don't asks if user wants to create
+    a new donor. Add a new donor to the public donors if desired.
+    If a donor existed, or the user wants to create a new donor then ask for the amount donated.
+    The donation is then added to the donor within the donors object and the object's generate_recent_thanks() function
+    is called.
+    """
+    current_donor = next((x for x in donors.donor_list if x.name.upper() == person.upper()), None)
+    if current_donor is None:
+        create_choice = input(person + " doesn't exist in the database, would you like to add them? (yes to add): ")
+        if create_choice.upper() == "YES":
+            donors.add_new_donor(person)
+            current_donor = donors.donor_list[-1]
+        else:
+            print(person + " was not added")
+    if current_donor is not None:
+        while True:  # Loop until a valid amount is provided
+            donation_amount = input("\nHow much did {} donate?: ".format(person))
+            try:
+                donation_amount = round(float(donation_amount), 2)
+            except ValueError:
+                print("Please provide a valid amount without $ symbol")
+            else:
+                if donation_amount > 0:
+                    break
+                else:
+                    print("Please provide a valid amount without $ symbol")
+        current_donor.new_donation(donation_amount)
+        print(current_donor.generate_recent_thanks())
```

Comment:
I was a bit confused reading the code in `thank_you` because it only called `write_letter` and couldn't find the code that adds / updates donors - this was because I expected `write_letter` to write a letter and nothing else.

This function ideally should be renamed or broken down into two functions that do appropriate things.